### PR TITLE
apiserver: use self-signed cert fixtures in integration test servers

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 1]
```diff
@@ -136,8 +139,37 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
...
+// GenerateSelfSignedCertKeyWithFixtures creates a self-signed certificate and key for the given host.
+// Host may be an IP or a DNS name. You may also specify additional subject alt names (either ip or dns names)
+// for the certificate.
+//
+// If fixtureDirectory is non-empty, it is a directory path which can contain pre-generated certs. The format is:
+// <host>_<ip>-<ip>_<alternateDNS>-<alternateDNS>.crt
+// <host>_<ip>-<ip>_<alternateDNS>-<alternateDNS>.key
+// Certs/keys not existing in that directory are created.
+func GenerateSelfSignedCertKeyWithFixtures(host string, alternateIPs []net.IP, alternateDNS []string, fixtureDirectory string) ([]byte, []byte, error) {
+	maxAge := time.Hour * 24 * 365 // one year self-signed certs
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/cert/cert.go]: Can we lengthen these? Would prefer to not have to regen annually

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 2]
```diff
@@ -149,7 +181,7 @@ func GenerateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS
 			CommonName: fmt.Sprintf("%s-ca@%d", host, time.Now().Unix()),
 		},
 		NotBefore: time.Now(),
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/cert/cert.go]: I'd suggest backdating these significantly as well, or clock skewed containerized envs will fail.

**Review by liggitt — COMMENTED** [review]
Two nits in valid date range

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: that's the default for the non-fixture case. 100y for the fixtures (below) should be enough.

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: good idea

Comment (sttts) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: done, substracted an hour from `notBefore`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt [hunk 3]
```diff
@@ -0,0 +1,37 @@
+-----BEGIN CERTIFICATE-----
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: do we have to do anything special to get bazel to pick these up (like name the folder testdata or something)?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: Very good point. Should be the same as in serving_test.go (if integration tests and unit tests work the same in bazel).

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: @mikedanese @ixdy what's the best way to do this? We need the testdata in all tests which (indirectly) import apiextensions-apiserver/pkg/cmd/server/testing (similar for kube-apiserver).

Comment (ixdy) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: If you name a directory `testdata`, I believe `gazelle` will include it as a `data` dependency on all `go_test`s in that package, though that isn't quite enough for what you want here, it sounds like?

The canonical bazel way would be to create a filegroup for those data files, and then explicitly list them as `data` dependencies in each test that needs them. `gazelle` can't figure out what packages those would be, however. 

You can try adding a `data` dependency on the relevant `go_library` itself (i.e. https://github.com/sttts/kubernetes/blob/3ad9a14ae479536ba8b62dc77e6981117f0fdeaa/cmd/kube-apiserver/app/testing/BUILD#L8), which might do the right thing, adding a transitive `data` dependency to any tests that depend on this library. I haven't tried it, but it's worth a shot.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: Integration tests are not bazel based, are they? So for the moment we should be fine it seems.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testfiles/localhost_127.0.0.1_localhost.crt]: Nevertheless, I added the testdata dependency to the `testing` go library target.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 4]
```diff
@@ -75,6 +78,11 @@ type GeneratableKeyCert struct {
 	// CertDirectory is a directory that will contain the certificates.  If the cert and key aren't specifically set
 	// this will be used to derive a match with the "pair-name"
 	CertDirectory string
+	// FixtureDirectory is a directory that contains test fixture used to avoid regeneration of certs during tests.
+	// The format is:
+	// <host>_<ip>-<ip>_<alternateDNS>-<alternateDNS>.crt
+	// <host>_<ip>-<ip>_<alternateDNS>-<alternateDNS>.key
```
Comment (jennybuckley) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: should it be ```<host>_<ip>_<ip>_<alternateDNS>_<alternateDNS>```?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: fixed

**Comment by sttts** [comment]
@jennybuckley @liggitt addressed comments and rebased. ptal.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: hrmm, just noticed this is adding test-only code into config structs... is this going to show up in external config types?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: These are not used externally in any config.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go [hunk 5]
```diff
@@ -109,11 +111,19 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
...
 
-	s.RecommendedOptions.SecureServing.Listener, s.RecommendedOptions.SecureServing.BindPort, err = createListenerOnFreePort()
+	s.RecommendedOptions.SecureServing.Listener, s.RecommendedOptions.SecureServing.BindPort, err = createLocalhostListenerOnFreePort()
 	if err != nil {
 		return result, fmt.Errorf("failed to create listener: %v", err)
 	}
 	s.RecommendedOptions.SecureServing.ServerCert.CertDirectory = result.TmpDir
+	s.RecommendedOptions.SecureServing.ExternalAddress = s.RecommendedOptions.SecureServing.Listener.Addr().(*net.TCPAddr).IP // use listener addr although it is a loopback device
+
+	_, thisFile, _, ok := runtime.Caller(0)
```
Comment (mikedanese) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: Is there an alternative to this? What does this assume about the working directory of the test?

**Comment by mikedanese** [comment]
What lead to this PR?

**Comment by liggitt** [comment]
> What led to this PR?

tests running in containers hanging because of lack of entropy needing to regenerate certs on every run

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: it is the positions of this file, i.e. the fixtures are next to it. This avoid that we need `n` fixtures for  `n` test packages using this test server.

**Pull request merged**